core/edr-core/src/main/java/org/eclipse/tractusx/edc/edr/core/service/EdrServiceImpl.java

```diff
@@ -74,12 +74,25 @@ private ServiceResult<DataAddress> autoRefresh(String id, DataAddress edr, Refre
         if (edrEntry == null) {
             return ServiceResult.notFound("An EndpointDataReferenceEntry with ID '%s' does not exist".formatted(id));
         }
-        if (isExpired(edr, edrEntry) || mode.equals(RefreshMode.FORCE_REFRESH)) {
-            monitor.debug("Token expired, need to refresh.");
-            return tokenRefreshHandler.refreshToken(id, edr)
-                    .compose(updated -> updateEdr(edrEntry, updated));
+        if (edrLock.isExpired(edr, edrEntry) || mode.equals(RefreshMode.FORCE_REFRESH)) {
+            var result = ServiceResult.from(edrLock.acquireLock(id, edr))
+                    .compose(shouldRefresh -> {
+                        if (!shouldRefresh) {
+                            monitor.debug("Dont need to refresh. Will resolve existing.");
+                            var refreshedEdr = edrStore.resolveByTransferProcess(id);
+                            return ServiceResult.from(refreshedEdr);
+                        } else {
+                            monitor.debug("Token '%s' expired, need to refresh.".formatted(id));
+                            return tokenRefreshHandler.refreshToken(id, edr)
+                                    .compose(updated -> updateEdr(edrEntry, updated));
+                        }
+                    });
+            edrLock.releaseLock(id);
```

Comment:
please handle the result somehow, e.g.
```suggestion
            edrLock.releaseLock(id).onFailure(f -> monitor.warning("Error releasing lock: %s".formatted(f.getFailureDetail())
```

Reply:
[42c1d78](https://github.com/eclipse-tractusx/tractusx-edc/pull/1633/commits/42c1d78d8ab47b6a92e845073a595c6fb42bc579)